Fix compiler crash and profitability test in memset optimisation

cme opened the pull request:
The legacy `getDestAlignment` member used to get alignment for existing `memset` intrinsics can use a value of 0 to represent undefined alignment, which will cause the alignment offset calculation to fail as it uses this as the denominator of a modulo operation.

Commit:
Fix compiler crash on undefined alignment memset

djtodoro commented on llvm/lib/Transforms/Scalar/MemCpyOptimizer.cpp:
```diff
@@ -265,7 +265,10 @@ void MemsetRanges::addRange(int64_t Start, int64_t Size, Value *Ptr,
+    if (I->MaxAlignmentOffset != 0)
+      I->MaxAlignmentOffset = (I->MaxAlignmentOffset + Size) % I->MaxAlignment;
+    else
+      I->MaxAlignmentOffset = 0;
```
Hmm, we are checking if the `MaxAlignmentOffset` is `0`, and if not we are initializing the variable, if yes we are again setting it to `0`.

cme replied:
Yes you're right this is redundant! Thanks

Commit:
Adapt memset heuristics and tests for targets with unaligned accesses

Targets with unaligned accesses can make a better job of inlining
memsets without knowing exact alignment.

Commit:
Fix denominator check

djtodoro commented on llvm/lib/Transforms/Scalar/MemCpyOptimizer.cpp:
```diff
@@ -297,7 +297,7 @@ void MemsetRanges::addRange(int64_t Start, int64_t Size, Value *Ptr,
     I->StartPtr = Ptr;
     I->Alignment = Alignment;
-    if (I->MaxAlignmentOffset != 0)
+    if (I->MaxAlignment != 0)
```
Thanks. This now makes sense.

djtodoro commented:
This makes sense to me.
This PR fixes:
```
  LLVM :: Transforms/MemCpyOpt/merge-into-memset.ll
  LLVM :: Transforms/MemCpyOpt/profitable-memset.ll
```

But, the new heuristics [0] introduced couple of other failures as well that still fail:

```
  Clang :: CodeGenCXX/auto-var-init.cpp
  LLVM :: Transforms/MemCpyOpt/form-memset.ll
```

[0] https://github.com/MediaTek-Labs/llvm-project/commit/efeed9e372fb8616c3b457789e84614b897ca16d

djtodoro commented:
@cme what is the status of this?
